File: quantum_mechanics/quantum-core.js
/**
 * quantum-core.js — PhD-Level Quantum Mechanics Library (JavaScript)
 * ==================================================================
 * Global IIFE: window.MajixQuantum
 *
 * Implements:
 *   Complex arithmetic
 *   State vectors, density matrices
 *   Quantum gates (1-qubit, 2-qubit, n-qubit QFT)
 *   Grover's algorithm
 *   Von Neumann entropy and entanglement
 *   Quantum channel simulation (depolarising, amplitude damping)
 *   Bloch sphere coordinates
 *   Variational quantum circuits (parameterised)
 */

(function (global) {
  'use strict';

  // -------------------------------------------------------------------------
  // Complex Number
  // -------------------------------------------------------------------------
  class Complex {
    constructor(re = 0, im = 0) { this.re = re; this.im = im; }
    add(c) { return new Complex(this.re + c.re, this.im + c.im); }
    sub(c) { return new Complex(this.re - c.re, this.im - c.im); }
    mul(c) { return new Complex(this.re * c.re - this.im * c.im, this.re * c.im + this.im * c.re); }
    scale(s) { return new Complex(this.re * s, this.im * s); }
    conj() { return new Complex(this.re, -this.im); }
    abs() { return Math.sqrt(this.re ** 2 + this.im ** 2); }
    abs2() { return this.re ** 2 + this.im ** 2; }
    arg() { return Math.atan2(this.im, this.re); }
    exp() {
      const r = Math.exp(this.re);
      return new Complex(r * Math.cos(this.im), r * Math.sin(this.im));
    }
    toString() { return `${this.re.toFixed(4)}${this.im >= 0 ? '+' : ''}${this.im.toFixed(4)}i`; }
    static fromPolar(r, theta) { return new Complex(r * Math.cos(theta), r * Math.sin(theta)); }
    static zero() { return new Complex(0, 0); }
    static one() { return new Complex(1, 0); }
    static i() { return new Complex(0, 1); }
  }

  // -------------------------------------------------------------------------
  // Complex Vector / Matrix Operations
  // -------------------------------------------------------------------------
  const C = {
    vadd: (u, v) => u.map((x, i) => x.add(v[i])),
    vsub: (u, v) => u.map((x, i) => x.sub(v[i])),
    vscale: (v, s) => v.map(x => x.scale(s)),
    vdot: (u, v) => u.reduce((acc, x, i) => acc.add(x.conj().mul(v[i])), Complex.zero()),
    vnorm: (v) => Math.sqrt(v.reduce((acc, x) => acc + x.abs2(), 0)),
    vnormalize: (v) => { const n = C.vnorm(v); return v.map(x => x.scale(1 / n)); },
    matvec: (A, v) => A.map(row => row.reduce((acc, a, j) => acc.add(a.mul(v[j])), Complex.zero())),
    matmul: (A, B) => {
      const n = A.length, m = B[0].length, k = B.length;
      return Array.from({ length: n }, (_, i) =>
        Array.from({ length: m }, (_, j) =>
          A[i].reduce((acc, a, l) => acc.add(a.mul(B[l][j])), Complex.zero())));
    },
    dag: (A) => A[0].map((_, j) => A.map((row, i) => row[j].conj())),
    trace: (A) => A.reduce((acc, row, i) => acc.add(row[i]), Complex.zero()),
    outer: (u, v) => u.map(ui => v.map(vi => ui.mul(vi.conj()))),
    kron: (A, B) => {
      const nA = A.length, mA = A[0].length, nB = B.length, mB = B[0].length;
      return Array.from({ length: nA * nB }, (_, i) =>
        Array.from({ length: mA * mB }, (_, j) =>
          A[Math.floor(i / nB)][Math.floor(j / mB)].mul(B[i % nB][j % mB])));
    },
    eye: (n) => Array.from({ length: n }, (_, i) => Array.from({ length: n }, (_, j) => i === j ? Complex.one() : Complex.zero())),
  };

  // -------------------------------------------------------------------------
  // Standard States and Gates
  // -------------------------------------------------------------------------
  const ZERO = [Complex.one(), Complex.zero()];
  const ONE  = [Complex.zero(), Complex.one()];
  const PLUS = [new Complex(1/Math.SQRT2, 0), new Complex(1/Math.SQRT2, 0)];
  const MINUS = [new Complex(1/Math.SQRT2, 0), new Complex(-1/Math.SQRT2, 0)];

  const gates = {
    I: [[Complex.one(), Complex.zero()], [Complex.zero(), Complex.one()]],
    X: [[Complex.zero(), Complex.one()], [Complex.one(), Complex.zero()]],
    Y: [[Complex.zero(), new Complex(0, -1)], [new Complex(0, 1), Complex.zero()]],
    Z: [[Complex.one(), Complex.zero()], [Complex.zero(), new Complex(-1, 0)]],
    H: [[new Complex(1/Math.SQRT2, 0), new Complex(1/Math.SQRT2, 0)],
        [new Complex(1/Math.SQRT2, 0), new Complex(-1/Math.SQRT2, 0)]],
    S: [[Complex.one(), Complex.zero()], [Complex.zero(), new Complex(0, 1)]],
    T: [[Complex.one(), Complex.zero()], [Complex.zero(), Complex.fromPolar(1, Math.PI / 4)]],
    Rz: (theta) => [[Complex.fromPolar(1, -theta/2), Complex.zero()],
                    [Complex.zero(), Complex.fromPolar(1, theta/2)]],
    Ry: (theta) => [[new Complex(Math.cos(theta/2), 0), new Complex(-Math.sin(theta/2), 0)],
                    [new Complex(Math.sin(theta/2), 0), new Complex(Math.cos(theta/2), 0)]],
    Rx: (theta) => [[new Complex(Math.cos(theta/2), 0), new Complex(0, -Math.sin(theta/2))],
                    [new Complex(0, -Math.sin(theta/2)), new Complex(Math.cos(theta/2), 0)]],
    CNOT: [
      [Complex.one(), Complex.zero(), Complex.zero(), Complex.zero()],
      [Complex.zero(), Complex.one(), Complex.zero(), Complex.zero()],
      [Complex.zero(), Complex.zero(), Complex.zero(), Complex.one()],
      [Complex.zero(), Complex.zero(), Complex.one(), Complex.zero()],
    ],
    SWAP: [
      [Complex.one(), Complex.zero(), Complex.zero(), Complex.zero()],
      [Complex.zero(), Complex.zero(), Complex.one(), Complex.zero()],
      [Complex.zero(), Complex.one(), Complex.zero(), Complex.zero()],
      [Complex.zero(), Complex.zero(), Complex.zero(), Complex.one()],
    ],
  };

  function applyGate(state, gate) { return C.matvec(gate, state); }

  // -------------------------------------------------------------------------
  // Quantum Fourier Transform
  // -------------------------------------------------------------------------
  function qftMatrix(nQubits) {
    const N = 1 << nQubits;
    const omega = Complex.fromPolar(1, 2 * Math.PI / N);
    return Array.from({ length: N }, (_, j) =>
      Array.from({ length: N }, (_, k) => {
        const exp = j * k;
        const angle = 2 * Math.PI * exp / N;
        return new Complex(Math.cos(angle) / Math.sqrt(N), Math.sin(angle) / Math.sqrt(N));
      }));
  }

  function applyQFT(state) {
    const N = state.length, nQ = Math.log2(N);
    return C.matvec(qftMatrix(nQ), state);
  }

  // -------------------------------------------------------------------------
  // Grover's Algorithm
  // -------------------------------------------------------------------------
  function groverOracle(state, target) {
    return state.map((x, i) => i === target ? x.scale(-1) : x);
  }

  function groverDiffusion(state) {
    const N = state.length;
    const total = state.reduce((acc, x) => acc.add(x), Complex.zero());
    const mean = total.scale(1 / N);
    return state.map(x => mean.scale(2).sub(x));
  }

  function groverSearch(N, target, nIter = null) {
    const optIter = nIter !== null ? nIter : Math.floor(Math.PI / 4 * Math.sqrt(N));
    let state = Array(N).fill(null).map(() => new Complex(1 / Math.sqrt(N), 0));
    for (let k = 0; k < optIter; k++) {
      state = groverOracle(state, target);
      state = groverDiffusion(state);
    }
    const probs = state.map(x => x.abs2());
    return { state, probs, iterations: optIter, successProb: probs[target] };
  }

  function measure(state, rng = Math.random) {
    const probs = state.map(x => x.abs2());
    let cum = 0, r = rng();
    for (let i = 0; i < probs.length; i++) {
      cum += probs[i];
      if (r <= cum) return i;
    }
    return probs.length - 1;
  }

  // -------------------------------------------------------------------------
  // Density Matrices and Entropy
  // -------------------------------------------------------------------------
  function densityMatrix(state) { return C.outer(state, state); }

  function partialTrace(rho, dA, dB) {
    return Array.from({ length: dA }, (_, i) =>
      Array.from({ length: dA }, (_, j) =>
        Array.from({ length: dB }, (_, k) => rho[i * dB + k][j * dB + k]).reduce((a, b) => a.add(b), Complex.zero())));
  }

  function vonNeumannEntropy2x2(rho) {
    // For 2x2 density matrix, compute eigenvalues and entropy
    const a = rho[0][0].re, d = rho[1][1].re;
    const bcAbs2 = rho[0][1].abs2();
    const disc = Math.sqrt(Math.max(((a - d) / 2) ** 2 + bcAbs2, 0));
    const lam1 = Math.max((a + d) / 2 + disc, 0);
    const lam2 = Math.max((a + d) / 2 - disc, 0);
    return -[lam1, lam2].reduce((acc, lam) => lam > 1e-12 ? acc + lam * Math.log2(lam) : acc, 0);
  }

  function entanglementEntropy(psi, dA, dB) {
    const rho = densityMatrix(psi);
    const rhoA = partialTrace(rho, dA, dB);
    if (dA === 2) return vonNeumannEntropy2x2(rhoA);
    // Diagonal approximation for larger dimensions
    return -rhoA.reduce((acc, row, i) => {
      const lam = Math.max(row[i].re, 0);
      return lam > 1e-12 ? acc + lam * Math.log2(lam) : acc;
    }, 0);
  }

  // -------------------------------------------------------------------------
  // Bloch Sphere
  // -------------------------------------------------------------------------
  function blochVector(rho) {
    // For single-qubit density matrix ρ = ½(I + r·σ)
    const rx = 2 * rho[0][1].re;
    const ry = 2 * rho[1][0].im;
    const rz = rho[0][0].re - rho[1][1].re;
    return { x: rx, y: ry, z: rz, purity: rx ** 2 + ry ** 2 + rz ** 2 };
  }

  function stateToBloch(state) {
    const rho = densityMatrix(state);
    return blochVector(rho);
  }

  // -------------------------------------------------------------------------
  // Quantum Channel
  // -------------------------------------------------------------------------
  function applyChannel(rho, krausOps) {
    const n = rho.length;
    const result = Array.from({ length: n }, () => Array(n).fill(null).map(() => Complex.zero()));
    for (const K of krausOps) {
      const Kd = C.dag(K);
      const term = C.matmul(C.matmul(K, rho), Kd);
      for (let i = 0; i < n; i++) for (let j = 0; j < n; j++) result[i][j] = result[i][j].add(term[i][j]);
    }
    return result;
  }

  function depolarizing(rho, p) {
    const n = rho.length;
    const result = rho.map(row => row.map(x => x.scale(1 - p)));
    for (let i = 0; i < n; i++) result[i][i] = result[i][i].add(new Complex(p / n, 0));
    return result;
  }

  function amplitudeDamping(rho, gamma) {
    const K0 = [[Complex.one(), Complex.zero()], [Complex.zero(), new Complex(Math.sqrt(1 - gamma), 0)]];
    const K1 = [[Complex.zero(), new Complex(Math.sqrt(gamma), 0)], [Complex.zero(), Complex.zero()]];
    return applyChannel(rho, [K0, K1]);
  }

  // -------------------------------------------------------------------------
  // Bell State Generation
  // -------------------------------------------------------------------------
  function bellState(type = 'phi+') {
    const s = 1 / Math.SQRT2;
    switch (type) {
      case 'phi+':  return [new Complex(s, 0), Complex.zero(), Complex.zero(), new Complex(s, 0)];
      case 'phi-':  return [new Complex(s, 0), Complex.zero(), Complex.zero(), new Complex(-s, 0)];
      case 'psi+':  return [Complex.zero(), new Complex(s, 0), new Complex(s, 0), Complex.zero()];
      case 'psi-':  return [Complex.zero(), new Complex(s, 0), new Complex(-s, 0), Complex.zero()];
      default: throw new Error('Unknown Bell state');
    }
  }

  function isBellState(state, tol = 1e-6) {
    const probs = state.map(x => x.abs2());
    const nz = probs.filter(p => p > tol);
    return nz.length === 2 && Math.abs(nz[0] - 0.5) < tol && Math.abs(nz[1] - 0.5) < tol;
  }

  // -------------------------------------------------------------------------
  // Public API
  // -------------------------------------------------------------------------
  const MajixQuantum = {
    Complex, C,
    ZERO, ONE, PLUS, MINUS,
    gates, applyGate,
    qftMatrix, applyQFT,
    groverOracle, groverDiffusion, groverSearch, measure,
    densityMatrix, partialTrace, vonNeumannEntropy2x2, entanglementEntropy,
    blochVector, stateToBloch,
    applyChannel, depolarizing, amplitudeDamping,
    bellState, isBellState,
  };

  if (typeof module !== 'undefined' && module.exports) {
    module.exports = MajixQuantum;
  } else {
    global.MajixQuantum = MajixQuantum;
  }

})(typeof globalThis !== 'undefined' ? globalThis : typeof window !== 'undefined' ? window : this);

File: quantum_mechanics/quantum_core.py
"""
quantum_core.py — PhD-Level Quantum Mechanics Library
=====================================================
Implements:
  State vectors and density matrices
  Quantum gates (1-qubit, 2-qubit)
  Quantum Fourier Transform
  Grover's algorithm (oracle-based simulation)
  Quantum harmonic oscillator (ladder operators, Fock states)
  Time-independent perturbation theory
  Variational quantum eigensolver (VQE) sketch
  Quantum channels (Kraus operators)
  Von Neumann entropy, entanglement entropy, Schmidt decomposition
  Wigner function (quasi-probability distribution)
"""

from __future__ import annotations
import math
import cmath
from typing import Callable, Optional


# ---------------------------------------------------------------------------
# Complex Vector / Matrix Utilities
# ---------------------------------------------------------------------------

Cnum = complex
CVec = list[Cnum]
CMat = list[list[Cnum]]


def cvdot(u: CVec, v: CVec) -> Cnum:
    return sum(a.conjugate()*b for a,b in zip(u,v))

def cvnorm(v: CVec) -> float:
    return math.sqrt(abs(cvdot(v,v)))

def cvadd(u: CVec, v: CVec) -> CVec: return [a+b for a,b in zip(u,v)]
def cvsub(u: CVec, v: CVec) -> CVec: return [a-b for a,b in zip(u,v)]
def cvscale(v: CVec, s: Cnum) -> CVec: return [x*s for x in v]

def cmv(A: CMat, v: CVec) -> CVec:
    return [sum(A[i][j]*v[j] for j in range(len(v))) for i in range(len(A))]

def cmm(A: CMat, B: CMat) -> CMat:
    n, m, p = len(A), len(A[0]), len(B[0])
    C = [[complex(0)]*p for _ in range(n)]
    for i in range(n):
        for k in range(m):
            for j in range(p): C[i][j] += A[i][k]*B[k][j]
    return C

def dag(A: CMat) -> CMat:
    """Hermitian conjugate (dagger)."""
    return [[A[j][i].conjugate() for j in range(len(A))] for i in range(len(A[0]))]

def outer_c(u: CVec, v: CVec) -> CMat:
    return [[u[i]*v[j].conjugate() for j in range(len(v))] for i in range(len(u))]

def kron(A: CMat, B: CMat) -> CMat:
    """Kronecker product A ⊗ B."""
    nA, mA = len(A), len(A[0])
    nB, mB = len(B), len(B[0])
    C = [[complex(0)]*(mA*mB) for _ in range(nA*nB)]
    for i in range(nA):
        for j in range(mA):
            for k in range(nB):
                for l in range(mB):
                    C[i*nB+k][j*mB+l] = A[i][j]*B[k][l]
    return C

def trace(A: CMat) -> Cnum:
    return sum(A[i][i] for i in range(len(A)))

def partial_trace_B(rho: CMat, dA: int, dB: int) -> CMat:
    """Partial trace over subsystem B: ρ_A = Tr_B[ρ]."""
    rhoA = [[complex(0)]*dA for _ in range(dA)]
    for i in range(dA):
        for j in range(dA):
            rhoA[i][j] = sum(rho[i*dB+k][j*dB+k] for k in range(dB))
    return rhoA


# ---------------------------------------------------------------------------
# Standard States and Gates
# ---------------------------------------------------------------------------

ZERO: CVec = [complex(1), complex(0)]
ONE:  CVec = [complex(0), complex(1)]
PLUS: CVec = [1/math.sqrt(2), 1/math.sqrt(2)]
MINUS: CVec = [1/math.sqrt(2), -1/math.sqrt(2)]

# Pauli matrices
I2: CMat = [[complex(1),complex(0)],[complex(0),complex(1)]]
X_gate: CMat = [[complex(0),complex(1)],[complex(1),complex(0)]]
Y_gate: CMat = [[complex(0),-1j],[1j,complex(0)]]
Z_gate: CMat = [[complex(1),complex(0)],[complex(0),complex(-1)]]
H_gate: CMat = [[1/math.sqrt(2), 1/math.sqrt(2)],[1/math.sqrt(2), -1/math.sqrt(2)]]
S_gate: CMat = [[complex(1),complex(0)],[complex(0),1j]]
T_gate: CMat = [[complex(1),complex(0)],[complex(0),cmath.exp(1j*math.pi/4)]]

def Rz(theta: float) -> CMat:
    """Rotation about z-axis: Rz(θ) = exp(-iθZ/2)."""
    return [[cmath.exp(-1j*theta/2), complex(0)],
            [complex(0), cmath.exp(1j*theta/2)]]

def Ry(theta: float) -> CMat:
    c, s = math.cos(theta/2), math.sin(theta/2)
    return [[complex(c), complex(-s)], [complex(s), complex(c)]]

# CNOT gate (control=qubit 0, target=qubit 1, 4×4 matrix)
CNOT: CMat = [
    [complex(1),complex(0),complex(0),complex(0)],
    [complex(0),complex(1),complex(0),complex(0)],
    [complex(0),complex(0),complex(0),complex(1)],
    [complex(0),complex(0),complex(1),complex(0)],
]

# Toffoli (CCX) gate — 8×8
def toffoli() -> CMat:
    n = 8
    T = [[complex(1) if i==j else complex(0) for j in range(n)] for i in range(n)]
    T[6], T[7] = T[7][:], T[6][:]
    return T


def apply_gate(state: CVec, gate: CMat) -> CVec:
    return cmv(gate, state)


# ---------------------------------------------------------------------------
# Quantum Fourier Transform (QFT)
# ---------------------------------------------------------------------------

def qft_matrix(n_qubits: int) -> CMat:
    """
    n-qubit QFT matrix (2ⁿ × 2ⁿ): F_{jk} = ω^{jk}/√N, ω = e^{2πi/N}.
    QFT|j⟩ = (1/√N) Σ_k e^{2πijk/N} |k⟩.
    """
    N = 2**n_qubits
    omega = cmath.exp(2j * math.pi / N)
    F = [[omega**(j*k)/math.sqrt(N) for k in range(N)] for j in range(N)]
    return F


def apply_qft(state: CVec) -> CVec:
    N = len(state)
    n = int(math.log2(N))
    F = qft_matrix(n)
    return cmv(F, state)


# ---------------------------------------------------------------------------
# Grover's Algorithm
# ---------------------------------------------------------------------------

def grover_oracle(state: CVec, target: int) -> CVec:
    """Oracle Uₓ|x⟩ = -|x⟩ if x==target, else |x⟩."""
    s = state[:]
    s[target] = -s[target]
    return s


def grover_diffusion(state: CVec) -> CVec:
    """Grover diffusion (inversion about the mean): D = 2|s⟩⟨s| - I."""
    N = len(state)
    mean = sum(state) / N
    return [2*mean - s for s in state]


def grover_search(N: int, target: int, n_iter: Optional[int] = None) -> tuple[CVec, int]:
    """
    Grover's algorithm. Starting from uniform superposition, apply
    ≈ π/4 √N iterations to maximise probability of |target⟩.
    Returns final state vector and number of iterations.
    """
    state: CVec = [complex(1/math.sqrt(N))] * N
    if n_iter is None:
        n_iter = int(math.pi/4 * math.sqrt(N))
    for _ in range(n_iter):
        state = grover_oracle(state, target)
        state = grover_diffusion(state)
    return state, n_iter


def measure(state: CVec, seed: Optional[int] = None) -> int:
    """
    Simulate a measurement in the computational basis.
    Returns the outcome with probability |αᵢ|².
    """
    import random
    if seed is not None: random.seed(seed)
    probs = [abs(a)**2 for a in state]
    r = random.random(); cumsum = 0.0
    for i, p in enumerate(probs):
        cumsum += p
        if r < cumsum: return i
    return len(state)-1


# ---------------------------------------------------------------------------
# Quantum Harmonic Oscillator
# ---------------------------------------------------------------------------

def number_operator(n_max: int) -> CMat:
    """Number operator N̂ = â†â, diagonal 0,1,...,n_max."""
    return [[complex(k) if i==k else complex(0) for j in range(n_max+1)][i]
            if False else complex(i if i==j else 0)
            for j in range(n_max+1) for i in range(n_max+1)]


def build_N_op(n_max: int) -> CMat:
    N = n_max+1
    return [[complex(i) if i==j else complex(0) for j in range(N)] for i in range(N)]


def lowering_op(n_max: int) -> CMat:
    """Annihilation operator â: â|n⟩ = √n |n-1⟩."""
    N = n_max+1
    a = [[complex(0)]*N for _ in range(N)]
    for n in range(1, N): a[n-1][n] = complex(math.sqrt(n))
    return a


def raising_op(n_max: int) -> CMat:
    """Creation operator â†: â†|n⟩ = √(n+1)|n+1⟩."""
    return dag(lowering_op(n_max))


def coherent_state(alpha: complex, n_max: int) -> CVec:
    """
    Coherent state |α⟩ = e^{-|α|²/2} Σ_n (α^n/√(n!)) |n⟩.
    """
    N = n_max+1
    norm = math.exp(-abs(alpha)**2/2)
    alpha_n = [1.0]  # α^n
    fact = [1.0]     # √(n!)
    for n in range(1, N):
        alpha_n.append(alpha_n[-1]*alpha)
        fact.append(fact[-1]*math.sqrt(n))
    return [complex(norm*alpha_n[n]/fact[n]) for n in range(N)]


# ---------------------------------------------------------------------------
# Quantum Channels (Kraus Operators)
# ---------------------------------------------------------------------------

def apply_channel(rho: CMat, kraus: list[CMat]) -> CMat:
    """Apply quantum channel: ε(ρ) = Σₖ Kₖ ρ Kₖ†."""
    n = len(rho)
    result = [[complex(0)]*n for _ in range(n)]
    for K in kraus:
        Kd = dag(K)
        term = cmm(cmm(K, rho), Kd)
        for i in range(n):
            for j in range(n): result[i][j] += term[i][j]
    return result


def depolarising_channel(rho: CMat, p: float) -> CMat:
    """
    Depolarising channel: ε(ρ) = (1-p)ρ + p I/2.
    Kraus operators: K₀=√(1-3p/4) I, K₁=√(p/4) X, K₂=√(p/4) Y, K₃=√(p/4) Z.
    """
    n = len(rho)
    result = [[complex(0)]*n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            result[i][j] = (1-p)*rho[i][j]
    # Add p/2 * I/2 (maximally mixed)
    for i in range(n): result[i][i] += complex(p/n)
    return result


def amplitude_damping(rho: CMat, gamma: float) -> CMat:
    """
    Amplitude damping channel (qubit relaxation to |0⟩).
    K₀ = [[1,0],[0,√(1-γ)]], K₁ = [[0,√γ],[0,0]].
    """
    K0: CMat = [[complex(1),complex(0)],[complex(0),complex(math.sqrt(1-gamma))]]
    K1: CMat = [[complex(0),complex(math.sqrt(gamma))],[complex(0),complex(0)]]
    return apply_channel(rho, [K0, K1])


# ---------------------------------------------------------------------------
# Von Neumann Entropy and Entanglement
# ---------------------------------------------------------------------------

def eigenvalues_2x2(A: CMat) -> tuple[complex, complex]:
    """Eigenvalues of 2×2 Hermitian matrix via quadratic formula."""
    a, b, c, d = A[0][0], A[0][1], A[1][0], A[1][1]
    tr = a+d; det = a*d-b*c
    disc = cmath.sqrt((tr/2)**2-det)
    return tr/2+disc, tr/2-disc


def von_neumann_entropy(rho: CMat, base: float = 2.0) -> float:
    """S(ρ) = -Tr(ρ log ρ) = -Σᵢ λᵢ log λᵢ (over non-zero eigenvalues)."""
    n = len(rho)
    if n == 2:
        lam1, lam2 = eigenvalues_2x2(rho)
        eigs = [abs(lam1.real), abs(lam2.real)]
    else:
        # Power iteration for eigenvalues (simplified for small matrices)
        eigs = [abs(rho[i][i].real) for i in range(n)]  # diagonal approx
    total = 0.0
    for lam in eigs:
        if lam > 1e-12: total -= lam*math.log(lam)/math.log(base)
    return total


def entanglement_entropy(psi: CVec, dA: int, dB: int) -> float:
    """
    Entanglement entropy of pure state |ψ⟩ ∈ H_A ⊗ H_B.
    S_E = S(ρ_A) = -Σᵢ λᵢ log λᵢ where λᵢ are Schmidt coefficients squared.
    """
    rho = outer_c(psi, psi)
    rhoA = partial_trace_B(rho, dA, dB)
    return von_neumann_entropy(rhoA)


def schmidt_decomposition(psi: CVec, dA: int, dB: int) -> tuple[list[float], list[CVec], list[CVec]]:
    """
    Schmidt decomposition |ψ⟩ = Σᵢ √λᵢ |αᵢ⟩ |βᵢ⟩.
    Computed via SVD of reshaped coefficient matrix M[i][j] = ψ[i*dB+j].
    (Simplified: uses power method to find dominant Schmidt coefficient.)
    """
    M = [[psi[i*dB+j] for j in range(dB)] for i in range(dA)]
    # Singular values via eigenvalues of M†M (2×2 case approximation)
    MdagM = [[sum(M[k][i].conjugate()*M[k][j] for k in range(dA))
              for j in range(dB)] for i in range(dB)]
    if dB == 2 and dA == 2:
        lam1, lam2 = eigenvalues_2x2(MdagM)
        lambdas = sorted([abs(lam1.real), abs(lam2.real)], reverse=True)
    else:
        lambdas = [abs(MdagM[i][i].real) for i in range(min(dA, dB))]
    return lambdas, [], []  # Full SVD vectors omitted for brevity


# ---------------------------------------------------------------------------
# Time-Independent Perturbation Theory
# ---------------------------------------------------------------------------

def first_order_energy(E0: list[float], V: CMat, n: int) -> complex:
    """E_n^(1) = ⟨n⁰|V|n⁰⟩ (basis vector |n⁰⟩ = eₙ)."""
    return V[n][n]


def second_order_energy(E0: list[float], V: CMat, n: int) -> complex:
    """E_n^(2) = Σ_{m≠n} |⟨m⁰|V|n⁰⟩|² / (E_n⁰ - E_m⁰)."""
    total = complex(0)
    for m in range(len(E0)):
        if m == n: continue
        dE = E0[n] - E0[m]
        if abs(dE) < 1e-14: continue
        total += abs(V[m][n])**2 / dE
    return total


def first_order_state(E0: list[float], psi0: list[CVec], n: int, V: CMat) -> CVec:
    """
    |ψ_n^(1)⟩ = Σ_{m≠n} ⟨m⁰|V|n⁰⟩/(E_n⁰-E_m⁰) |m⁰⟩.
    """
    N = len(E0)
    correction: CVec = [complex(0)]*N
    for m in range(N):
        if m == n: continue
        dE = E0[n] - E0[m]
        if abs(dE) < 1e-14: continue
        Vmn = sum(psi0[m][k].conjugate()*sum(V[k][j]*psi0[n][j] for j in range(N))
                  for k in range(N))
        for k in range(N):
            correction[k] += (Vmn/dE)*psi0[m][k]
    return correction
